### PR TITLE
track progress from within segment processor framework

### PR DESCRIPTION
**Pull request by klsince**
Follow up on #9449 to get finer grained progress status from SegmentProcessorFramework as well, as that's usually the step takes most of task time, breaking it down to map/reduce/generation steps for now.

### CODE REVIEW

**Comment by klsince** [comment]
e.g. for those progress status from SegmentProcessorFramework: 
```
{
      "ts": 1663957811428,
      "status": "Doing map phase on data from RecordReader: 1/3"
    },
    {
      "ts": 1663957811434,
      "status": "Doing map phase on data from RecordReader: 2/3"
    },
    {
      "ts": 1663957811434,
      "status": "Doing map phase on data from RecordReader: 3/3"
    },
    {
      "ts": 1663957811434,
      "status": "Doing reduce phase on data from partition: 0, 1/1"
    },
    {
      "ts": 1663957811440,
      "status": "Creating segment of sequentId: 0 with data from partition: 0 in row range: [0, 3)/3"
    },
```

File: pinot-minion/src/main/java/org/apache/pinot/minion/taskfactory/TaskFactoryRegistry.java [hunk 1]
```diff
@@ -127,17 +127,17 @@ private TaskResult runInternal() {
...
                 _minionMetrics.addMeteredTableValue(taskType, MinionMeter.NUMBER_TASKS_FATAL_FAILED, 1L);
                 LOGGER.error("Caught fatal exception while executing task: {}", _taskConfig.getId(), e);
-                return new TaskResult(TaskResult.Status.FATAL_FAILED, e.toString());
+                return new TaskResult(TaskResult.Status.FATAL_FAILED, StringUtil.getStackTraceAsString(e));
               } catch (Exception e) {
                 _eventObserver.notifyTaskError(pinotTaskConfig, e);
                 _minionMetrics.addMeteredTableValue(taskType, MinionMeter.NUMBER_TASKS_FAILED, 1L);
                 LOGGER.error("Caught exception while executing task: {}", _taskConfig.getId(), e);
-                return new TaskResult(TaskResult.Status.FAILED, e.toString());
+                return new TaskResult(TaskResult.Status.FAILED, StringUtil.getStackTraceAsString(e));
```
Comment (klsince) [inline pinot-minion/src/main/java/org/apache/pinot/minion/taskfactory/TaskFactoryRegistry.java:140]: e.toString() like e.getMessage() just returns the error msg (basically the first line of the stack trace).

File: pinot-core/src/main/java/org/apache/pinot/core/segment/processing/mapper/SegmentMapper.java [hunk 2]
```diff
@@ -105,8 +108,12 @@ public SegmentMapper(List<RecordReader> recordReaders, SegmentProcessorConfig pr
    */
   public Map<String, GenericRowFileManager> map()
       throws Exception {
+    Consumer<Object> observer = _processorConfig.getProgressObserver();
+    int totalCount = _recordReaders.size();
+    int count = 1;
     GenericRow reuse = new GenericRow();
     for (RecordReader recordReader : _recordReaders) {
+      observer.accept(String.format("Doing map phase on data from RecordReader: %d/%d", count++, totalCount));
```
Comment (zhtaoxiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/processing/mapper/SegmentMapper.java:116]: It seems to me that it's hard to get the meaning of those two numbers. Maybe change it to `RecordReader %d (%d in total)` or something similar?

Comment (klsince) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/processing/mapper/SegmentMapper.java:116]: sounds good!

**Review by zhtaoxiang — APPROVED** [review]
LGTM

**Review by jackjlli — APPROVED** [review]
LGTM

**Pull request merged**